teps/0015-implicit-mapping.md

```diff
@@ -0,0 +1,658 @@
+
+```yaml
+    - name: deploy
+      runAfter: [verify]
+      taskRef:
+        name: install-tekton-release
+```
+
+### Goal
+- Simplify the `Pipeline` specification and therefore the `Pipeline` YAML by:
+  - Allowing a `Pipeline` author to omit parameters and resources from `PipelineTasks` that match in name with parameters and resources listed under the `Pipeline.spec.params` and `Pipeline.spec.resources` fields.
+
+* Support finally `Tasks` for parameters and resources (no support for using results from previous `Tasks`)
+
+### Non Goals
+- Support implicit mapping for parameters and resources in `Conditions`.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+|#|Requirement                                                                                                                                                                                                                                                                                                                                                                          |
+|-|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+|1| A `Pipeline` author can omit the explicit specification of parameters and resources for some, or any, `Task` in the `Pipeline.spec.tasks` field if:<br><br> <ul><li> The pipeline is annotated for implicit mapping <i><br>AND<br></i> <li> The omitted parameters and resources can be resolved according to Req. 2 </li></ul>|
+|2| Tekton tries to resolve omitted parameters and resources for a pipeline `Task` in the following order: <br><br> <ol><li> If any preceding `Task` has a `Task` result whose name matches an implicit parameter name of the `Task`, resolve the implicit parameter value with the last produced `Task` result that matches. <li> If a parameter or resource name in the `Pipeline.spec` field matches a `Task's` implicit parameter or resource name, resolve the parameter and/or resource values. </li> <li> Tekton produces an error if parameters or resources can't be matched. <br> *\*Explicitly specified params and resources always take precedence.*
+
+[DISCUSSION]: Additionally, omitting input resources that use the output resource of a previous task by using the [`from`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-the-from-parameter) parameter could be supported.
+This could mean making the `from` parameter optional.
+When omitted, Tekton can match an input resource to the latest matching output resource.
+
+## Proposal
+
+Allow for an implicit specification of a `Pipeline` wherein parameters and resources for a `PipelineTask` can be omitted.
+
+<!-- #### User Story
```

Comment:
Maybe the User Story could be:

> As a Pipeline author I want the parameters and resources in a Pipeline Spec to be automatically handed to Tasks that share the parameter or resource names so that my Pipelines are easier to consume and it becomes harder to make mistakes editing the YAML in future.